swarm.js

```diff
@@ -1,38 +1,146 @@
 'use strict'
 const peerInfo = require('./lib/peer-info')
+const peerQueue = require('./lib/queue')
 const { EventEmitter } = require('events')
 const guts = require('@hyperswarm/guts')
 
+const MAX_PEERS_DEFAULT = 16
+const ERR_MISSING_KEY = 'key is required and must be a buffer'
+
 module.exports = opts => new Swarm(opts)
 
 class Swarm extends EventEmitter {
   constructor (opts = {}) {
     super()
-    const { bootstrap, ephemeral } = opts
-    this.network = guts({
+    const {
+      maxPeers = MAX_PEERS_DEFAULT,
+      bootstrap,
+      ephemeral
+    } = opts
+    const queue = peerQueue()
+    const network = guts({
       bootstrap,
       ephemeral,
-      bind: () => {
-        this.emit('listening')
-      },
+      bind: () => this.emit('listening'),
       socket: (socket, isTCP) => {
         const info = peerInfo(null)
         info.connected(socket, isTCP)
         this.emit('connection', socket, info)
       },
-      close: () => {
-        this.emit('close')
-      }
+      close: () => this.emit('close')
     })
+    queue.on('readable', this._drain(queue))
+    this.peers = 0
+    this.maxPeers = maxPeers
+    this.emphemeral = ephemeral !== false
+    this.network = network
+    this.queue = queue
+  }
+  _drain (queue) {
+    const onConnect = (info) => (err, socket, isTCP) => {
+      if (err) {
+        this.peers -= 1
+        if (!queue.requeue(info)) queue.requeue(info)
+        drain()
+        return
+      }
+      if (info.reconnecting) info.reconnect(false)
```

Comment:
why do we disable reconnects here?

Reply:
I might have misunderstood, but I thought `reconnect(bool)` set whether a peer was reconnecting or not, 
if we've successfully connected then we need to set reconnecting to false

Reply:
Ah yea I can see the confusion. The RECONNECT flag is used as a capability internally. I.e. should this peer be reconnecting at all? The `info.reconnect(bool)` is the public method to control this.

Something like

```js
swarm.on('connection', function (connection, peerInfo) {
  // do stuff
  if (weDontWantToReconnectToThisPeerWhenItCloses) {
    peerInfo.reconnect(false)
  }
})
```